### PR TITLE
Added AIX LV module

### PR DESCRIPTION
**Pull request by adejoux**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME
aix_lvol.py

##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 9958b215d7) last updated 2017/01/23 17:32:04 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

This module is adding support to create, change and delete AIX logical volumes. I started from **lvol** module but AIX LVM management is really different. 

Here a simple playbook using it: 

```
- hosts: adxlpar2
  gather_facts: False
  tasks:
    - name: Create a logical volume of 512M.
      aix_lvol: vg=testvg lv=testlv size=512M
    - name: Create a logical volume of 512M with disks hdisk1 and hdisk2
      aix_lvol: vg=testvg lv=test2lv size=512M pvs="hdisk1,hdisk2"
    - name: Create a logical volume of 512M mirrored.
      aix_lvol: vg=testvg lv=test3lv size=512M copies=2
    - name: Create a logical volume of 1G with a minimum placement policy.
      aix_lvol: vg=testvg lv=test4lv size=1G policy=minimum
    - name: Change logical volume size to 1200 MB.
      aix_lvol: vg=testvg lv=test4lv size=1200M
    - name: delete logical volumes
      aix_lvol: vg=testvg lv={{ item }} state="absent"
      with_items:
        - "testlv"
        - "test2lv"
        - "test3lv"
        - "test4lv"
```

And the output:

```
ansible-playbook  -i hosts playbook.yml

PLAY [adxlpar2] ************************************************************************************************************************************************************

TASK [Create a logical volume of 512M.] ************************************************************************************************************************************
changed: [adxlpar2]

TASK [Create a logical volume of 512M with disks hdisk1 and hdisk2] ********************************************************************************************************
changed: [adxlpar2]

TASK [Create a logical volume of 512M mirrored.] ***************************************************************************************************************************
changed: [adxlpar2]

TASK [Create a logical volume of 1G with a minimum placement policy.] ******************************************************************************************************
changed: [adxlpar2]

TASK [Change logical volume size to 1200 MB.] ******************************************************************************************************************************
changed: [adxlpar2]

TASK [delete logical volumes] **********************************************************************************************************************************************
changed: [adxlpar2] => (item=testlv)
changed: [adxlpar2] => (item=test2lv)
changed: [adxlpar2] => (item=test3lv)
changed: [adxlpar2] => (item=test4lv)

PLAY RECAP *****************************************************************************************************************************************************************
adxlpar2                   : ok=6    changed=6    unreachable=0    failed=0  
```

### CODE REVIEW

File: lib/ansible/modules/system/aix_lvol.py
Comment (ryansb) [inline lib/ansible/modules/system/aix_lvol.py]: This doesn't seem to add `--test`, is the above comment out of date, or is this option name wrong?

File: lib/ansible/modules/system/aix_lvol.py
Comment (ryansb) [inline lib/ansible/modules/system/aix_lvol.py]: Instead of setting this out here and doing the `exit_json` out of scope, can you please use `exit_json` right in the various `rc==0` blocks?

File: lib/ansible/modules/system/aix_lvol.py
Comment (ryansb) [inline lib/ansible/modules/system/aix_lvol.py]: These docs should also include a `RETURNS` section so users can see examples of what the module outputs in different scenarios.

**Review by ryansb — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/aix_lvol.py
Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: The comment is out of date. I will update.

File: lib/ansible/modules/system/aix_lvol.py
Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: Ok. I will update.

**Comment by molekuul** [comment]
I have a question on your code:
In the function "convert_size "
you have two times the
```
if  match.group(1) == "G":
```
but 1 time with the size_factor = 1024
and 1 time with size_factor = 1024*1024
see snippet below 
this doesn't seem right to me?

```
def convert_size(size):
    base_size = None
    size_factor = None

    match = re.search("(\d+)", size)
    if match is not None:
        base_size = int(match.group(1))

    match = re.search("(\w)", size)
    if match is not None:

        if match.group(1) == "G":
            size_factor=1024
        if match.group(1) == "G":
            size_factor=1024*1024

    if base_size is None:
```

kind regards,

Joris Weijters

**Comment by adejoux** [comment]
@molekuul, thanks it was a mistake on my side. It's for terabytes sizing. I didn't tetsed this one. I fixed it.

**Review by ryansb — APPROVED** [review]

**Comment by molekuul** [comment]
shipit

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [inline lib/ansible/modules/system/aix_lvol.py]: IMO, adding `.upper()` make the parse more robust.
```python
 if match.group(1).upper() == "T":
```

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [inline lib/ansible/modules/system/aix_lvol.py]: You can use `type='list'` with `default=list()`.
Later, you can use it like `' '.join(pvs)`.

https://github.com/natefoo/ansible/blob/599eb6115e53fd6fd15770a4c7ef292177d5b888/lib/ansible/module_utils/basic.py#L1525

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [inline lib/ansible/modules/system/aix_lvol.py]: Why not using `type='int' default=1`?

Comment (tmshn) [reply lib/ansible/modules/system/aix_lvol.py]: Also, you can do
```python
lv_type=dict(default='jfs2'),
# ...snip...
opts=dict(default=''),
```

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [inline lib/ansible/modules/system/aix_lvol.py]: What are these 2 lines doing??

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [inline lib/ansible/modules/system/aix_lvol.py]: You don't need `test_opt` here?

**Review by tmshn — CHANGES_REQUESTED** [review]

**Comment by adejoux** [comment]
@tmshn thanks for your comments. I will implement your recommendations tonight.

**Review by tmshn — APPROVED** [review]

**Comment by tmshn** [comment]
Looks good, shipit.

File: lib/ansible/modules/system/aix_lvol.py
Comment (mscherer) [inline lib/ansible/modules/system/aix_lvol.py]: type should be a 'int'

File: lib/ansible/modules/system/aix_lvol.py
Comment (mscherer) [inline lib/ansible/modules/system/aix_lvol.py]: It would IMHO make more sense to move that near the other command, so we can verify both command at the same time.

File: lib/ansible/modules/system/aix_lvol.py
Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: Hello, Thanks for the comment.
 If you are talking about the "size" attribute, I made it a string to be able to take in parameter "100m", "100g" or "1T". I think it's making it easier to use for the end user.

File: lib/ansible/modules/system/aix_lvol.py
Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: I moved lslv_cmd. I was thinking about moving the others "_cmd" commands but deletion and creation commands will never be run together in the same run so I am not sure. Let me know if you want me to move them too.

File: lib/ansible/modules/system/aix_lvol.py
Comment (tmshn) [reply lib/ansible/modules/system/aix_lvol.py]: No, maybe @mscherer is talking about `copies`.

Do
```
copies=dict(default=1, type='int')
```
rather than
```
copies=dict(default='1')
```

Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: Thanks. I modified it. :)

**Comment by adejoux** [comment]
Hello,

I am not familiar with the process. Should I do something to be merged ?

Regards,

Alain

**Comment by tmshn** [comment]
bot_status

**Comment by adejoux** [comment]
ready_for_review

**Comment by adejoux** [comment]
@tmshn thanks again :)

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: One way to do this is something like:

```yaml
    unit = size[-1]
    units = list('b', 'k', 'M', 'G', 'T', 'P', 'E', 'Z', 'Y')
    try:
        multiplier = 1024**units.index(unit)
    except ValueError::
        e = get_exception()
        module.fail_json(msg="No valid size unit specified. %s" % e)
```

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: If parameters are not required, you do not have to add `required: false`. This is the implied value, so it is easier to see what is required if only the required parameters have this option.

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Please put all imports together at the top, and include only the stuff you are using, i.e.
```python
from ansible.module_utils.basic import AnsibleModule
```

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: If it is an integer, don't add the quotes because that would make it a string.

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Please add `type="str"`, and maybe list the possibilities as `choices=`.

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Please add `type="str"`

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Please add `type="str"`

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Please add `type="str"`

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: We are not doing anything with stdout and stderr, but on failures it may be the only clue to why it failed.

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: We are not doing anything with stdout and stderr, but on failures it may be the only clue to why it failed.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/system/aix_lvol.py
Comment (adejoux) [reply lib/ansible/modules/system/aix_lvol.py]: Like it :) I will implement something like that along your others recommendations. thanks

**Comment by adejoux** [comment]
bot_status

**Comment by adejoux** [comment]
Hello @dagwieers , Are you waiting on something from me?

Thanks

**Review by dagwieers — APPROVED** [review]

**Comment by adejoux** [comment]
bot_status

**Comment by adejoux** [comment]
Hello @dagwieers ,@tmshn, @mscherer,

Not sure what is blocking for the merge now. Could you help me ?

Thanks,

Alain

**Comment by tmshn** [comment]
It looks every time someone request changes, the `maintainer_shipits` will be reset to 0, and maintainers should approve it by `shipit` comment every time `needs_revision` changed to `False`.
It's a bad behavior of ansibot 😥 ...though I don't know if it's a bug or intended.

Now it looks all green ✅ ; so you can ask for the `shipit` comment again for formerly commented maintainers.

**Comment by tmshn** [comment]
shipit

**Comment by dagwieers** [comment]
@tmshn The idea is that changes after a review/shipit needs to be re-reviewed. Because a shipit on old code may not be relevant to the latest changes. So this is to be expected. Obviously, if the changes are rather small, this is a nuisance...

File: lib/ansible/modules/system/aix_lvol.py
Comment (dagwieers) [inline lib/ansible/modules/system/aix_lvol.py]: Just noticed a typo in the email-address.

**Comment by dagwieers** [comment]
I personally don't like the check_mode implementation, but confirm it's hard to do right in this case.

shipit

**Comment by jctanner** [comment]
@tmshn the count resets after a new commit. If someone requests changes, the "needs revision" label is set, but the "ship it" is not changed.

**Comment by tmshn** [comment]
@jctanner Ah I see. Actually I see it in action just now! lol
And that sounds appropriate; if something has changed, the changes must pass reviews.
Thanks for the info!

**Comment by tmshn** [comment]
shipit, finally.

**Comment by adejoux** [comment]
Hello @dagwieers ,@tmshn, 

I didn't expect than fixing a small typo  in the email address would make me do so many small fixes but at least I think it's ready now. And cleaner ;)

So could you approve it a last time. :)

Thanks,

Alain

**Review by dagwieers — APPROVED** [review]
shipit

**Comment by dagwieers** [comment]
bot_status

**Comment by dagwieers** [comment]
shipit

@jctanner Maybe a nice feature for ansibot: accept shipit statements from approved review process.

**Comment by tmshn** [comment]
shipit

**Pull request merged**
